hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -1417,6 +1430,70 @@ public void rename(final Path source, final Path destination,
     } while (shouldContinue);
   }
 
+  private void renameBlobDir(final Path source,
+      final Path destination,
+      final TracingContext tracingContext,
+      final ListBlobQueue listBlobQueue,
+      final BlobProperty blobPropOnSrc, final AbfsBlobLease srcDirBlobLease,
+      final Boolean isAtomicRename) throws AzureBlobFileSystemException {
+    BlobList blobList;
+    ListBlobConsumer listBlobConsumer = new ListBlobConsumer(listBlobQueue);
+    while(!listBlobConsumer.isCompleted()) {
+      blobList = listBlobConsumer.consume();
+      if(blobList == null) {
+        continue;
+      }
+      List<Future> futures = new ArrayList<>();
+      for (BlobProperty blobProperty : blobList.getBlobPropertyList()) {
+        futures.add(renameBlobExecutorService.submit(() -> {
+          try {
+            AbfsBlobLease blobLease = null;
+            if (isAtomicRename) {
+              /*
+               * Conditionally get a lease on the source blob to prevent other writers
+               * from changing it. This is used for correctness in HBase when log files
+               * are renamed. It generally should do no harm other than take a little
+               * more time for other rename scenarios. When the HBase master renames a
+               * log file folder, the lease locks out other writers.  This
+               * prevents a region server that the master thinks is dead, but is still
+               * alive, from committing additional updates.  This is different than
+               * when HBase runs on HDFS, where the region server recovers the lease
+               * on a log file, to gain exclusive access to it, before it splits it.
+               */
+              blobLease = new AbfsBlobLease(client,
+                  blobProperty.getPath().toUri().getPath(), tracingContext);
+            }
+            if(srcDirBlobLease != null) {
+              srcDirBlobLease.renewIfRequired();
+            }
+            renameBlob(
+                createDestinationPathForBlobPartOfRenameSrcDir(destination,
+                    blobProperty, source),
+                tracingContext, blobProperty.getPath(), blobLease != null ? blobLease.getLeaseId() : null);
+          } catch (AzureBlobFileSystemException e) {
+            LOG.error(String.format("rename from %s to %s for blob %s failed",
+                source, destination, blobProperty.getPath()), e);
+            throw new RuntimeException(e);
```

Comment:
Question: Why are we throwing Runtime Exception and not ABFSException itself??

Reply:
The method implemented here is for Runnable. Runnable's signature doesn't declare any exception in signature. Only RuntimeException can be thrown as its an unchecked exception.